### ISSUE TITLE
GLTFExporter caches textures not images

### ISSUE THREAD

**Issue by kickblade** [mrdoob/three.js#14063]
The current implementation of GLTFExporter is caching textures as it processes the scene but other parts of three js (e.g. TextureLoader and ImageLoader) cache the image.

This creates problems when, for instance, you load various model files (via FBXLoader etc) that have references to the same image files and you try to export the resulting scene via GLTFExporter with embedImages = true and binary = true. Essentially, it packs duplicate image data into the glb, which bloats it considerably.

The problem seems to centre around the fact that TextureLoader creates a new texture (with a potentially cached Image) every time you call load() but GLTFExporter is basing it's image caching on the texture when you try to process the scene. You end up with a list of textures all referring to the same image but GLTFExporter will go through each texture and write out the corresponding cached image.

NOTE: I could solve this by hacking in something to patch/reformat all the textures in my scene to work with GLTFExporter but I think handling this in the exporter would be better and more user-friendly.

##### Three.js version

- [ ] Dev
- [x ] r92
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

**Comment by donmccurdy** [comment]
Agreed, this should be handled in the exporter. Ideally we should cache both textures and images, independently.

**Comment by takahirox** [comment]
Yup, we have a plan to do that, but just haven't done yet. PR is welcome.

**Comment by kickblade** [comment]
OK cool, thanks for the feedback guys. I'll see if I can help out with this, I have a pretty good test case to work with. Will be in touch. And great work on supporting GLTF btw, finding it really useful.

**Comment by Usnul** [comment]
One extra possibility is to compute hashes for texture data (binary) and cache based on those. There's a possibility of collision, but that is pretty much negligible for SHA-256.

TL;DR
Cache based on data, not URL

**Comment by donmccurdy** [comment]
I think for now a `Map` based on the `texture.image` instance is sufficient — there are other complications on the PR anyway. I believe `texture.clone()` preserves `.image`.

**Comment by takahirox** [comment]
IMO, the current `Map` approach is reasonable. Detecting the same image with hash is a bit optimization rather than exporting. Such optimizations can be done out of the exporter, for example in the user side.
